crates/turbopack-core/src/chunk/mod.rs

```diff
@@ -168,59 +168,167 @@ impl ChunkGroupVc {
+/// recursively. The order of the result is non-deterministic, we also keep
+/// track of children indices, so we may re-build the original graph with the
+/// right order in [`build_adjacency_map`].
+///
+/// Also returns the total number of chunks in the graph.
+async fn get_chunks_edges_unsorted(root: ChunkVc) -> Result<(Vec<ChunkEdge>, usize)> {
+    let mut visited = HashSet::new();
+    let edges = [
+        // Start with an edge from the root chunk to itself. This edge will be skipped later,
+        // its only purpose is to start the traversal.
+        ChunkEdge {
+            parent: root,
+            child: root,
+            child_idx: 0,
+        },
+    ]
+    .into_iter()
+    .try_flat_map_recursive_join({
+        |edge| {
+            if visited.insert(edge.child) {
+                Some(get_chunk_children(edge.child))
+            } else {
+                None
+            }
+        }
+    })
+    .await?;
+    let chunks_count = visited.len();
 
-        Ok(chunks)
+    Ok((edges, chunks_count))
+}
+
+/// Builds the adjacency map of the chunk graph.
+fn build_adjacency_map(
+    edges: Vec<ChunkEdge>,
+    chunks_count: usize,
+) -> HashMap<ChunkVc, BTreeMap<usize, ChunkVc>> {
+    let mut adjacency_map = HashMap::with_capacity(chunks_count);
+    for edge in edges
+        .iter()
+        // Skip the root chunk.
+        .skip(1)
+    {
+        adjacency_map
+            .entry(edge.parent)
+            .or_insert_with(|| BTreeMap::new())
+            .insert(edge.child_idx, edge.child);
+    }
+    adjacency_map
+}
+
+/// Runs a post-order DFS pass on the chunk DAG to compute a reverse topological
+/// order.
+fn reverse_topological_order(
+    root: ChunkVc,
+    adjacency_map: HashMap<ChunkVc, BTreeMap<usize, ChunkVc>>,
+    chunks_count: usize,
+) -> Vec<ChunkVc> {
+    enum Pass {
+        Pre,
+        Post,
     }
+    let mut stack = vec![(Pass::Pre, root)];
+    let mut rev_topo_sort = Vec::with_capacity(chunks_count);
+    let mut visited = HashSet::new();
+    while let Some((pass, chunk)) = stack.pop() {
+        match pass {
+            Pass::Pre => {
+                if !visited.insert(chunk) {
+                    continue;
+                }
+
+                if let Some(children) = adjacency_map.get(&chunk) {
+                    stack.push((Pass::Post, chunk));
+                    stack.extend(children.values().copied().map(|child| (Pass::Pre, child)));
+                } else {
+                    // Run the post-order pass directly.
+                    rev_topo_sort.push(chunk);
+                }
+            }
+            Pass::Post => {
+                rev_topo_sort.push(chunk);
+            }
+        }
+    }
+
+    // To get a valid topological order, we would then reverse the vector.
+    // However, in this case, a reverse topological order is what we want.
+
+    rev_topo_sort
 }
 
 #[turbo_tasks::value_impl]
```

Comment:
<i class='graphite__hidden'>[Re: lines 537 to 537]<br/></i>

I had originally implemented a post-order DFS here instead of the reverse topological sort above, which fixed some issues but not all. It might still be relevant to also switch this to a reverse topological sort. @sokra ?<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/vercel/turbo/3771?utm_source=unchanged-line-comment">Graphite</a>.</span>

Reply:
What's the difference?